### PR TITLE
Add geometric distribution to new Generator

### PR DESCRIPTION
**Pull request by povinsahu1909**

### CODE REVIEW

File: cupy/random/_generator_api.pyx
Comment (emcastillo) [inline cupy/random/_generator_api.pyx:278]: p can be an array of floats
check numpy

https://numpy.org/doc/stable/reference/random/generated/numpy.random.Generator.geometric.html#numpy.random.Generator.geometric

File: tests/cupy_tests/random_tests/common_distributions.py
Comment (emcastillo) [inline tests/cupy_tests/random_tests/common_distributions.py:276]: test also arrays please

**Review by emcastillo — COMMENTED** [review]
test in https://github.com/cupy/cupy/blob/master/tests/cupy_tests/random_tests/test_generator.py is not refactored, should be like

https://github.com/cupy/cupy/blob/master/tests/cupy_tests/random_tests/test_generator.py#L576-L580

File: tests/cupy_tests/random_tests/common_distributions.py
Comment (povinsahu1909) [reply tests/cupy_tests/random_tests/common_distributions.py:276]: I tried passing the array as test parameters, but as of now we have to explicitly pass p as cupy.ndarray for distributions of Generator, but in RandomState there is no such requirement. And tests fails due to broadcast error.

Comment (takagi) [reply tests/cupy_tests/random_tests/common_distributions.py:276]: > And tests fails due to broadcast error.

Maybe the shape of `p` and `size` make a conflict?

File: cupy/random/_generator_api.pyx
Comment (takagi) [inline cupy/random/_generator_api.pyx:309]:
```suggestion
        y = ndarray(size if size is not None else (), numpy.int64)
```

**Comment by takagi** [comment]
I think this PR is basically fine except for some points left. Would you make a performance comparison against SciPy? Also, I'd like to merge this in this week if possible as we have GSoC's first evaluation next week. It's clearly plus for you if you make a merged PR before it.

File: tests/cupy_tests/random_tests/common_distributions.py
Comment (povinsahu1909) [reply tests/cupy_tests/random_tests/common_distributions.py:276]: https://github.com/cupy/cupy/blob/d53212bb87579013ec8a1c4b8d31de8512f92b40/tests/cupy_tests/random_tests/common_distributions.py#L111-L116

Is there any specific se of defining the numpy_len = 1000? 
As this is causing the broadcast error. I replaced the above line with following and it works well for the p as ndarray, but was failing for the `lognormal` and `normal` by rejecting null hypothesis.

```python
# kwargs['size'] = numpy_len
dtype = kwargs.pop('dtype', None)
numpy_rng = self.get_rng(numpy, self.__seed)
func_numpy = getattr(numpy_rng, self.target_method)(*args, **kwargs)
vals_numpy = func_numpy
count = 1 + (numpy_len - 1) // vals_numpy.size
for _ in range(1, count):
    vals_numpy = numpy.append(vals_numpy, func_numpy)

if dtype is not None:
    vals_numpy = vals_numpy.astype(dtype, copy=False)
vals_numpy = numpy.stack(vals_numpy).ravel()
```

Note: This change was only for testing the distribution of RandomState as for Generator `p` is required to be cupy.ndarray or a scalar.

**Comment by povinsahu1909** [comment]
Following is the performance comparision result measured on GTX1050
size(rows * columns) | scipy (sec) | cupy (sec)
--- | --- | ---
10000 * 1 | 0.001 | 0.000
12000 * 1 | 0.001 | 0.000
15000 * 1 | 0.001 | 0.000
20000 * 1 | 0.002 | 0.000
10000 * 1000 | 0.826 | 0.001
12000 * 1000 | 0.996 | 0.001
15000 * 1000 | 1.299 | 0.001
20000 * 1000 | 1.752 | 0.001
10000 * 2000 | 1.668 | 0.001
12000 * 2000 | 2.001 | 0.001
15000 * 2000 | 2.695 | 0.002
20000 * 2000 | 3.445 | 0.001
10000 * 3000 | 2.596 | 0.001
12000 * 3000 | 3.127 | 0.001
15000 * 3000 | 3.968 | 0.002
20000 * 3000 | 5.273 | 0.001

<details>
<summary>script</summary>


import numpy as np
import cupy as cp
import cupyx

n_repeat = 2
n_warmup = 1
crng = cp.random.default_rng()
nrng = np.random.default_rng()

def get_time(perf):
    cpu_time = perf.cpu_times.mean()
    gpu_time = perf.gpu_times.mean()
    return max(cpu_time, gpu_time)

def run(p, q):
    times = []
    perf = cupyx.time.repeat(nrng.geometric,
                             (p,),
                             n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    perf = cupyx.time.repeat(crng.geometric,
                             (q,),
                             n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    return times

print('size(nrows * ncols) | scipy (sec) | cupyx (sec)')
print('---|---|---')
nrows = [10000, 12000, 15000, 20000]
ncols = [1, 1000, 2000, 3000]
for m in ncols:
    for n in nrows:
        p = np.random.random((n, m))
        q = cp.asarray(p)
        times = run(p, q)
        print('{} x {} | {:.3f} | {:.3f}'.format(n, m, times[0], times[1]))

</details>

File: cupy/random/_generator_api.pyx
Comment (povinsahu1909) [inline cupy/random/_generator_api.pyx:303]: Can we remove this test, as due to this test input array is required to be cupy.ndarray. So something like `rng.geometric([0.2, 0.3, 0.4])` will not work, but was supported by numpy.

**Comment by takagi** [comment]
pfnCI, test this please.

**Comment by takagi** [comment]
Seems that cupy generates wrong numbers with larger size more than 1d. I get a similar result with another distribution, i.g. standard_normal, so it's supposed to be a general problem of the generator class.

```Python
import numpy as np
import cupy as cp
n, m = 1000, 1000
p = cp.asarray(np.random.random((n, m)))
rng = cp.random.default_rng()
print(rng.geometric(p))
```

```
[[ 1  3  3 ...  1  2  1]
 [51  1  2 ...  5  1  1]
 [ 1  1  2 ...  1  1  4]
 ...
 [ 0  0  0 ...  0  0  0]
 [ 0  0  0 ...  0  0  0]
 [ 0  0  0 ...  0  0  0]]
```

**Comment by takagi** [comment]
Would you try the performance comparison again with flattened shapes `m*n` with the same conditions?

File: cupy/random/_generator_api.pyx
Comment (takagi) [reply cupy/random/_generator_api.pyx:303]: Making p take array-like other than cupy.ndarray is okay.

Comment (emcastillo) [reply cupy/random/_generator_api.pyx:303]: we only allow cupy arrays or scalars in order to avoid implicit synchronization.

**Comment by povinsahu1909** [comment]
Performance comparison for flattened shapes
size | numpy | cupy
--- | --- | ---
10000 | 0.001 | 0.000
12000 | 0.001 | 0.000
15000 | 0.001 | 0.000
20000 | 0.002 | 0.000
10000000 | 0.803 | 0.025
12000000 | 0.981 | 0.030
15000000 | 1.205 | 0.038
20000000 | 1.603 | 0.051
24000000 | 1.918 | 0.076
30000000 | 2.411 | 0.080
36000000 | 2.941 | 0.092
40000000 | 3.310 | 0.105
45000000 | 3.678 | 0.135
60000000 | 4.931 | 0.153

<details>
<summary>script</summary>
import numpy as np
import cupy as cp
import cupyx

n_repeat = 2
n_warmup = 1
crng = cp.random.default_rng()
nrng = np.random.default_rng()

def get_time(perf):
    cpu_time = perf.cpu_times.mean()
    gpu_time = perf.gpu_times.mean()
    return max(cpu_time, gpu_time)

def run(p, q):
    times = []
    perf = cupyx.time.repeat(nrng.geometric, (p,), n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    perf = cupyx.time.repeat(crng.geometric, (q,), n_repeat=n_repeat, n_warmup=n_warmup)
    times.append(get_time(perf))
    return times

print('size(nrows * ncols) | scipy (sec) | cupyx (sec)')
print('---|---|---')
nrows = [10000, 12000, 15000, 20000]
ncols = [1, 1000, 2000, 3000]
for m in ncols:
    for n in nrows:
        p = np.random.random((n*m, ))
        q = cp.asarray(p)
        times = run(p, q)
        print('{} | {:.3f} | {:.3f}'.format(n*m, times[0], times[1]))

</details>

**Comment by takagi** [comment]
> Performance comparison for flattened shapes

Thanks! This result looks reasonable to me.

File: tests/cupy_tests/random_tests/common_distributions.py
Comment (takagi) [reply tests/cupy_tests/random_tests/common_distributions.py:276]: I don't think `check_ks` for `p` of ndarray makes much sense, so we can skip it.

**Comment by takagi** [comment]
pfnCI, test this please.

**Review by takagi — APPROVED** [review]
LGTM! Waiting for CI pass.

**Comment by takagi** [comment]
Thanks, now it is merged!

**Pull request merged**
